backupccl: prevent a new cluster from resuming an in progress backup job

msbutler opened the pull request:
Issue https://github.com/cockroachdb/cockroach/issues/112114 described a rare timeline in which a restored or cutover tenant
can resume a backup job that began on another tenant, which could lead to the
new tenant overwriting backup metadata.

This patch asserts that there does not exist an existing manifest checkpoint
file written by a different cluster. Note that this check already exists for
the resumption path in which the job details were already fully populated. So,
this patch merely adds the same check if on Resume, the coordinator does not
observe all job details in the job record.

Informs https://github.com/cockroachdb/cockroach/issues/112114

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/112113)
<!-- Reviewable:end -->

msbutler committed:
backupccl: prevent a new cluster from resuming an in progress backup job

Issue #112114 described a rare timeline in which a restored or cutover tenant
can resume a backup job that began on another tenant, which could lead to the
new tenant overwriting backup metadata.

This patch asserts that there does not exist an existing manifest checkpoint
file written by a different cluster. Note that this check already exists for
the resumption path in which the job details were already fully populated. So,
this patch merely adds the same check if on Resume, the coordinator does not
observe all job details in the job record.

Informs #112114

Release note: None

dt commented on pkg/ccl/backupccl/backup_job.go:
```diff
@@ -988,6 +988,39 @@ func maybeRelocateJobExecution(
+	kmsEnv cloud.KMSEnv,
+	user username.SQLUsername,
+) error {
+	defaultConf, err := cloud.ExternalStorageConfFromURI(details.URI, user)
```
Can we just do `if c := job.Payload().CreationClusterID; c != uuid.Nil && c != jobExec.ExtendedEvalContext().ClusterID { return errors.Newf(...)` rather than introducing more external IO?

dt commented on pkg/ccl/backupccl/backup_job.go:
```diff
@@ -932,6 +936,28 @@ func (b *backupResumer) Resume(ctx context.Context, execCtx interface{}) error {
+// this backup since resuming the backup may overwrite backup metadata created by a different cluster.
+func (b *backupResumer) ensureClusterIDMatches(ctx context.Context, clusterID uuid.UUID) error {
+	if createdBy := b.job.Payload().CreationClusterID; createdBy == uuid.Nil {
+		// This cluster was upgraded from a version that did not set a clusterID
```
We've been populating this for more than a release now, so I'd just remove this untested branch and just make the mismatch condition be id != nil && id != current.

msbutler replied:
Done.

msbutler commented:
bors r=dt

dt commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -3377,6 +3377,53 @@ func TestBackupTenantsWithRevisionHistory(t *testing.T) {
+	rng, _ := randutil.NewTestRand()
+	pausePoint := potentialPausePoints[rng.Intn(2)]
+	t10Sql.Exec(t, fmt.Sprintf("SET CLUSTER SETTING jobs.debug.pausepoints = '%s'", pausePoint))
+	_, err = t10.SQLConn(t, "").Exec("BACKUP INTO 'nodelocal://1/example-schedule'")
```
I think you could use `t10sql.ExpectErr`

msbutler replied:
done.

dt commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -3377,6 +3377,53 @@ func TestBackupTenantsWithRevisionHistory(t *testing.T) {
+	t10Sql.QueryRow(t, "WITH jobs AS (SHOW JOBS) SELECT job_id FROM jobs WHERE job_type='BACKUP';").Scan(&backupJobID)
+	jobutils.WaitForJobToPause(t, t10Sql, jobspb.JobID(backupJobID))
+
+	systemDB.Exec(t, `BACKUP TENANT 10 TO 'nodelocal://1/foo'`)
```
deprecated TO/FROM in case you care.

msbutler replied:
ah yeah. good point. done.

dt commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -3377,6 +3377,53 @@ func TestBackupTenantsWithRevisionHistory(t *testing.T) {
+			JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals()},
+	})
+	require.NoError(t, err)
+	t10Sql := sqlutils.MakeSQLRunner(t10.SQLConn(t, ""))
```
Looks like you need to rebase on top of #112086.

msbutler commented:
bors r=dt
